Add script for converting PromQL tests to new syntax format

kapillamba4 opened the pull request:
Fix for https://github.com/prometheus/prometheus/issues/16367

Adding a simple script to convert tests automatically to new syntax.
- Migrating to new syntax is idempotent operation.
- Multiple Modes supported

kapillamba4 commented:
Adding @beorn7 and @NeerajGartia21

beorn7 reviewed (commented):
Thank you very much.

I still have to look at the code in detail. (Maybe @NeerajGartia21 will beat me to that. :)

However, one note: I don't think we should put this as a 1st class top level binary into the directory structure. This is just for a one time transition. I would put the file with the Go `main` package somewhere underneath the `promqltest` directory and give instructions in the README.md how to `go run ...` it.

About the actual conversion in this PR: This should be done in a separate commit from the conversion script. Maybe even in a separate PR. And we should do a manual pass through it to see where the "strict" or the "basic" conversion makes more sense. (Bonus points for separate commits: First the auto-conversion, presumably in strict mode, and then a separate commit to manually relax the checks where it makes sense.)

beorn7 commented on promql/promqltest/testdata/aggregators.test:
```diff
@@ -235,21 +235,25 @@ eval instant at 0m max(http_requests)
+eval instant at 0m max(http_requests_histogram)
+  expect info
 
 eval instant at 0m min(http_requests)
```
Since we are explicitly interested in this _not_ creating an annotation, an `expect no_info` would make sense here.

beorn7 commented on promql/promqltest/testdata/at_modifier.test:
```diff
@@ -91,7 +91,8 @@ eval instant at 25s metric{job="1"} @ 50 + metric{job="1"} @ 100
 # Note that this triggers an info annotation because we are rate'ing a metric that does not end in `_total`.
-eval_info instant at 25s rate(metric{job="1"}[100s] @ 100) + label_replace(rate(metric{job="2"}[123s] @ 200), "job", "1", "", "")
+eval instant at 25s rate(metric{job="1"}[100s] @ 100) + label_replace(rate(metric{job="2"}[123s] @ 200), "job", "1", "", "")
+  expect info
```
I would leave out this line and the comment above (line 93). We essentially only had to add this because `eval` was originally too strict about annotations. Now we can remove both again.

beorn7 reviewed (changes requested):
As a follow-up to my previous comment, here one example each for both cases:
1. We want to convert a simple `eval` into a new-style test with `expect no_info` (or `expect no_warn` in other cases) in cases where the absence of an annotation is relevant.
2. We want to convert an old-style `eval_info` into a simple new-style `eval` in cases where we do not really care about annotations.

All the tests have to be judged similarly, as suggested in my previous comment. (It doesn't have to be rocket science, but it should generally make sense.)

beorn7 commented on promql/promqltest/test.go:
```diff
@@ -1651,3 +1653,107 @@ func timeMilliseconds(t time.Time) int64 {
+
+	basicAnnotations := map[string][]string{
+		"eval_fail":    {"  expect fail"},
+		"eval_warn":    {"  expect warn"},
+		"eval_info":    {"  expect info"},
```
To be really relaxed, these two should actually be empty (the new `eval` tolerates annotations).

Maybe we should have three modes: `tolerant` (converting `eval_warn` and `eval_info` to a simple `eval`), `basic` (as implemented here), and `strict` (as implemented above).

beorn7 commented on promql/promqltest/test.go:
```diff
@@ -1651,3 +1653,107 @@ func timeMilliseconds(t time.Time) int64 {
+	strictAnnotations := map[string][]string{
+		"eval_fail":    {"  expect fail", "  expect no_warn", "  expect no_info"},
+		"eval_warn":    {"  expect no_fail", "  expect warn", "  expect no_info"},
+		"eval_info":    {"  expect info", "  expect no_warn", "  expect no_info"},
```
```suggestion
		"eval_info":    {"  expect info", "  expect no_warn"},
```

beorn7 reviewed (changes requested):
Some preliminary code comments.

NeerajGartia21 commented on promql/promqltest/test.go:
```diff
@@ -1651,3 +1653,107 @@ func timeMilliseconds(t time.Time) int64 {
+
+	strictAnnotations := map[string][]string{
+		"eval_fail":    {"  expect fail", "  expect no_warn", "  expect no_info"},
+		"eval_warn":    {"  expect no_fail", "  expect warn", "  expect no_info"},
```
Also, I think in strict mode, adding expect no_warn and expect no_info lines does not make sense for eval_fail.

beorn7 replied:
Who knows, maybe one day we'll get errors with annotations.

NeerajGartia21 commented on promql/promqltest/test.go:
```diff
@@ -1651,3 +1653,107 @@ func timeMilliseconds(t time.Time) int64 {
+	for i := 0; i < len(lines); i++ {
+		matches := evalRegex.FindStringSubmatch(strings.TrimSpace(lines[i]))
+		if matches == nil {
+			result = append(result, lines[i])
```
That's a good point. I had an idea that might complicate this PR a bit, though. What if we go a step further and make all the test lines more structured? For example, we could have no tabs for lines like `load` or `eval`, and use a single tab consistently for data lines and expect lines. This could help with consistency in the long run.

beorn7 replied:
I don't think we need to go as far as mandating a style for tests.

But the conversion script should create self-consistent output.

kapillamba4 commented:
hey @beorn7 @NeerajGartia21 

I am right now hardcoding the directory path: "promql/promqltest/testdata" do we need to make it configurable?
Saw this comment: https://github.com/prometheus/prometheus/issues/15794#issuecomment-2632407046

We have 3 modes now: tolerant, basic, strict and the indentation per eval block is consistent (using reference from existing indentation) so that the things don;t look weird

The main method is also now under promqltest folder with instructions in the corresponding README.md

kapillamba4 commented:
> Thank you very much.
> 
> I still have to look at the code in detail. (Maybe @NeerajGartia21 will beat me to that. :)
> 
> However, one note: I don't think we should put this as a 1st class top level binary into the directory structure. This is just for a one time transition. I would put the file with the Go `main` package somewhere underneath the `promqltest` directory and give instructions in the README.md how to `go run ...` it.
> 
> About the actual conversion in this PR: This should be done in a separate commit from the conversion script. Maybe even in a separate PR. And we should do a manual pass through it to see where the "strict" or the "basic" conversion makes more sense. (Bonus points for separate commits: First the auto-conversion, presumably in strict mode, and then a separate commit to manually relax the checks where it makes sense.)

MigrateTestData is now part of new file- test_migrate.go with a doc comment on top

beorn7 commented:
Let's first finish the review here and then think about the conversion PR. (I guess we should have one PR with two commits. First the auto-conversion and then the manual adjustments.)

beorn7 commented:
I'll try to look at the code here in detail ASAP.

beorn7 commented:
> I am right now hardcoding the directory path: "promql/promqltest/testdata", do we need to make it configurable? Saw this comment: [#15794 (comment)](https://github.com/prometheus/prometheus/issues/15794#issuecomment-2632407046)

Indeed, the directory should be customizable by the user. This is mainly meant for converting tests outside of this repo. (For a one-time conversion just within this repo, we did not need to publish the script here as part of the repo.)

NeerajGartia21 reviewed (commented):
Overall, the code looks good. Apologies for the delayed review.

Additionally, it would be helpful to include tests for this script. This will make it easier for reviewers to verify the functionality without running it locally and will also benefit users by providing clear examples for how to use and test the script.

kapillamba4 commented on promql/promqltest/test_migrate.go:
```diff
@@ -0,0 +1,186 @@
+
+		if skipBlock {
+			result = append(result, lines[i])
+			i++
```
I've made some changes to simplify the loop logic

NeerajGartia21 commented on promql/promqltest/test_migrate.go:
```diff
@@ -0,0 +1,200 @@
+			continue
+		}
+
+		// Get leading whitespace from startLine using indentRegex
```
```suggestion
		// Get leading whitespace from startLine using indentRegex.
```
Just a nit.

NeerajGartia21 reviewed (approved):
Looks good to me! Just a small nit.
Thanks for the script @kapillamba4 !
Let's now wait for @beorn7's review.

beorn7 commented:
Thanks everyone. And apologies for the review details. I'm still caught between several events at work and in my usual life, but I will finally be back at reviewing code next week.

kapillamba4 commented on promql/promqltest/README.md:
```diff
@@ -164,3 +164,17 @@ There can be multiple `<expect>` lines for a given `<type>`. Each `<type>` valid
 If at least one `<expect>` line of type `warn` or `info` is present, then all corresponding annotations must have a matching `expect` line.
+
+#### Migrating Test Files to the New Syntax
+- All `.test` files in test directory will be updated in place.
```
Updated

kapillamba4 commented on promql/promqltest/README.md:
```diff
@@ -164,3 +164,17 @@ There can be multiple `<expect>` lines for a given `<type>`. Each `<type>` valid
+The --mode flag controls how annotation rules are applied during migration:
+- strict: Adds all recommended expect lines (e.g., `expect no_warn`, `expect no_info`).
+- basic: Only adds minimal expect lines for each command.
+- tolerant: Only adds `expect fail` and `expect ordered` wherever needed.
```
Updated

kapillamba4 commented on promql/promqltest/README.md:
```diff
@@ -164,3 +164,17 @@ There can be multiple `<expect>` lines for a given `<type>`. Each `<type>` valid
 
 If at least one `<expect>` line of type `warn` or `info` is present, then all corresponding annotations must have a matching `expect` line.
+
+#### Migrating Test Files to the New Syntax
```
Updated

kapillamba4 commented:
Thanks for reviewing @beorn7, I’ve implemented the doc updates you suggested. 

Have also rebased the [strict migration mode PR](https://github.com/prometheus/prometheus/pull/16585) with this PR

beorn7 reviewed (approved):
Great, thank you very much.
